src/Enzo/enzo_EnzoMethodSinkMaker.cpp

```diff
@@ -0,0 +1,583 @@
+{
+  return std::numeric_limits<double>::max();
+}
+
+//-------------------------------------------------------------------------------------------
+
+bool EnzoMethodSinkMaker::jeans_length_not_resolved_(Block * block, int i,
+						     double const_G) throw()
+{
+  Field field = block->data()->field();
+
+  // First, let's get the square of the sound speed.
+  // Do this by getting the specific internal energy in this cell, and then computing
+  // the sound speed squared from that.
+
+  // Get pointers to values of density field and specific total energy field
+  enzo_float * density      = (enzo_float*) field.values("density");
+  enzo_float * vx           = (enzo_float *) field.values("velocity_x");
+  enzo_float * vy           = (enzo_float *) field.values("velocity_y");
+  enzo_float * vz           = (enzo_float *) field.values("velocity_z");
+  enzo_float * specific_te  = (enzo_float *) field.values("total_energy");
+
+  // Not sure if this is the best way to do this
+  // EnzoConfig has three different "dual energy" attributes.
+  // One of them is `method_hydro_dual_energy` but it seems that EnzoMethodHydro is not
+  // actually used.
+  // Assume that if either of the other two are true, there an "specific internal energy" field.
+  bool dual_energy =
+    enzo::config()->ppm_dual_energy ||  enzo::config()->method_vlct_dual_energy;
+  const enzo_float * specific_ie_field
+    = dual_energy ? (enzo_float*) field.values("internal_energy") : nullptr;
+
+  // Get pointers to magnetic field values if they exist.
+  enzo_float * bx =
+    field.is_field("bfield_x") ? (enzo_float*) field.values("b_field_x") : nullptr;
+  enzo_float * by =
+    field.is_field("bfield_y") ? (enzo_float*) field.values("b_field_y") : nullptr;
+  enzo_float * bz =
+    field.is_field("bfield_z") ? (enzo_float*) field.values("b_field_z") : nullptr;
+
+  // Get the specific internal energy
+  enzo_float specific_ie_cell;
+  if (dual_energy) specific_ie_cell = specific_ie_field[i];
+  else if (bx)
+    specific_ie_cell = specific_te[i] -
+      0.5 * ( vx[i] * vx[i] + vy[i] * vy[i] + vz[i] * vz[i] ) -
+      0.5 * ( bx[i] * bx[i] + by[i] * by[i] + bz[i] * bz[i] ) / density[i];
+  else specific_ie_cell = specific_te[i] -
+	 0.5 * ( vx[i] * vx[i] + vy[i] * vy[i] + vz[i] * vz[i] );
+
+  // Now compute the square of the sound speed
+  const double gamma = enzo::config()->field_gamma;
+  const double c_s_2 =  gamma * (gamma - 1.0) * specific_ie_cell;
+
+  // Now compute the Jeans length
+  const enzo_float jeans_length = sqrt(cello::pi * c_s_2 / (const_G * density[i]));
+
+  // Get the maximum cell width
+  double hx, hy, hz;
+  block->cell_width(&hx, &hy, &hz);
+  const double max_cell_width = std::max(hx,std::max(hy,hz));
+
+  return (jeans_length < jeans_length_resolution_cells_ * max_cell_width);
+}
+
+//-----------------------------------------------------------------------------------------
+
+bool EnzoMethodSinkMaker::flow_is_converging_(Block * block,
```

Comment:
This is a really nice function. I wonder should this be made more global so that others can use this too? Calculating the velocity divergence crops up quite a lot?